Default to using touchend for Handler.Feature, not touchstart

davea opened the pull request:
This is an existing PR rebased on our `master` branch: https://github.com/openlayers/ol2/pull/1118 - no other changes made. [MS I made some other changes]

Compiled OL ready for FMS testing in this PR: https://github.com/mysociety/fixmystreet/pull/4395

ejn committed:
Add options to Control.SelectFeature to be able to influence the
behaviour of the associated Handler.Feature and Handler.Box (e.g.
changing whether the Handler.Feature swallows particular events).

ejn committed:
Add option to Handler.Feature to react on touchend instead of touchstart

ejn committed:
Select features on touchend instead of touchstart by default

davea commented:
Unfortunately this now means that if you attempt to pan the map and your touch starts on a marker nothing will happen.

You have to tap on an empty bit of the map, which is problematic in really busy areas, e.g.:

![Screenshot 2023-04-20 at 16 00 18](https://user-images.githubusercontent.com/4776/233406851-4915f22b-d3cf-46b1-9edf-6b55bcd9783e.png)

I briefly investigated how to allow panning in that situation but didn't get very far. I wonder if we could do something to make the tap targets smaller on mobile, e.g. removing the shadows or shrinking the markers a little bit...

dracos reviewed (approved):
Not sure what the first commit is actually doing, the change doesn't require the user of the featureOptions/boxOptions, looks like it might be for the tests I guess? Anyway, if it fixes the issue, woo!

dracos commented:
"Unfortunately this now means that if you attempt to pan the map and your touch starts on a marker nothing will happen." - sigh

dracos commented:
So we want to allow the pan to carry on wherever, what we don't want is the selection of the object to happen if it's a pan.

dracos commented:
Weird, how I read the code is that the dragging should be unaffected, only the selection. Wonder what's going on.

davea commented:
I mean it’s still better than the current behaviour of a random report appearing if you start on a marker 🙂

dracos commented:
Yes, agreed it is better! I'll take a look and see if I can spot anything obvious.

dracos commented on lib/OpenLayers/Handler/Feature.js:
```diff
@@ -132,11 +140,51 @@ OpenLayers.Handler.Feature = OpenLayers.Class(OpenLayers.Handler, {
-        this.startTouch(); 
+    	// do we callback on the touchstart or on the touchend event?
+    	if (this.callbackOnTouchend) {
+    		this.EVENTMAP.touchstart = this.EVENTMAP.mousedown;
```
EVENTMAP is global, I think, so this is changing it for every Feature Handler, presumably depending upon which one is last set up.

dracos replied:
have pushed a fix for this little thing I think

dracos committed:
Prevent editing of the global EVENTMAP object.

Does not change behaviour.

dracos committed:
Handler.Feature propagate mousedown/touchstarts.

The previous commits fix the issue where a touchstart was causing the
feature handler to fire, which then could confuse other handlers that
thought the touch was still happening (leading to multi-touch events,
e.g. the map zooming when you tried to pan it). However, dragging the
map starting on a feature then did nothing. We flip the flag to start
propagating these events, which allows dragging on features to work.

davea commented on lib/OpenLayers/Handler/Feature.js:
```diff
@@ -94,7 +102,7 @@ OpenLayers.Handler.Feature = OpenLayers.Class(OpenLayers.Handler, {
      *      whatever the value of stopDown. Defaults to true.
      */
-    stopDown: true,
+    stopDown: false,
```
For the sake of completeness I guess the comment on L103 should reflect the new default
